File: python/instance.py
#!/usr/bin/python

import numpy as np
import pandas as pd

class Instance:

    """
    A class used to represent an Instance of a classification problem

    ...

    Attributes
    ----------
    dataset : pandas dataframe
        a dataframe containing all the features for all the variables
    prediction_features : list of str
        list of all the features used for the prediction
    target_feature : str
        the target feature to predict
    atom_indexes : dictionary (int,int) -> int
        index of an atom given its value
    atoms : list of (int,int)
        list of atom values: variable index plus discrete value

    _positive_samples : list of int
        list of indexes of the positive samples in the datafram
    _negative_samples : list of int
        todo

    Methods
    -------
    compute_score()
        Compute the positive and negative score of each atom
    """

    #---------------------------------------------------------------------------
    def __init__(self):

        self.atom_indexes = None # dictionary of atom indexes
        self.atoms = None # list of atoms: variable index and discrete value
        self.atom_score = None # positive and negative score for each atom

        self.dataset = None
        self.prediction_features = None
        self.target_feature = None
        self.n_values = None # number of discrete values for each features

        self._pos_samples = None # positive samples of the instance
        self._neg_samples = None # negative samples of the instance

        return


    #---------------------------------------------------------------------------
    def _init_instance(self):

        # separate the instance into two subsets: positive and negative samples
        positive_dataset = self.dataset.loc[self.dataset[self.target_feature] == True]
        negative_dataset = self.dataset.loc[self.dataset[self.target_feature] == False]

        # list of sample labels for each class
        self._pos_samples = positive_dataset.index.copy()
        self._neg_samples = negative_dataset.index.copy()

        # size of each class
        n_positive = positive_dataset.shape[0]
        n_negative = negative_dataset.shape[0]

        self.n_values = {} # discrete values for each feature

        atom_index = 0

        # iterate over all prediction features of the dataset
        for feature in self.prediction_features:

            # count unique values
            pos_unique, pos_counts = np.unique(positive_dataset[feature], return_counts=True)
            pos_occ = dict(zip(pos_unique, pos_counts))

            neg_unique, neg_counts = np.unique(negative_dataset[feature], return_counts=True)
            neg_occ = dict(zip(neg_unique, neg_counts))

            # assemble and sort discrete values of the feature
            discrete_values = np.concatenate((pos_unique, neg_unique))
            discrete_values = np.sort(np.unique(discrete_values))

            self.n_values[feature] = len(discrete_values)

            # print('max val: ', np.max(discrete_values))

            for value in range(np.max(discrete_values)+1):

                self.atoms.append((feature, value))
                self.atom_indexes[self.atoms[-1]] = atom_index

                # compute the score:
                pos_score = n_positive
                if value in pos_unique:
                    pos_score -= pos_occ[value]

                # compute negative score
                neg_score = n_negative
                if value in neg_unique:
                    neg_score -= neg_occ[value]

                self.atom_score.append([pos_score, neg_score])

                atom_index += 1

            # create an atom and fill the score for each value
            # for value in discrete_values:
            #     self.atoms.append((feature, value))
            #     self.atom_indexes[self.atoms[-1]] = atom_index
            #
            #     # compute the score:
            #     pos_score = n_positive
            #     if value in pos_unique:
            #         pos_score -= pos_occ[value]
            #
            #     # compute negative score
            #     neg_score = n_negative
            #     if value in neg_unique:
            #         neg_score -= neg_occ[value]
            #
            #     self.atom_score.append([pos_score, neg_score])
            #
            #     atom_index += 1

        self.atom_score = np.array(self.atom_score) # convert to numpy array

        del positive_dataset
        del negative_dataset

        return


    #---------------------------------------------------------------------------
    def _init_sub_instance(self, pos_excluded, neg_excluded = []):

        # init the scores of a sub instance by modifying the original score
        # separate the istance into two subsets: positive and negative samples

        ex_pos_dataset = None
        ex_neg_dataset = None

        sub_pos = len(pos_excluded) > 0
        sub_neg = len(neg_excluded) > 0

        if sub_pos:
            ex_pos_dataset = self.dataset.loc[pos_excluded]
            self.atom_score[:,0] -= len(pos_excluded)

        if sub_neg:
            ex_neg_dataset = self.dataset.loc[neg_excluded]
            self.atom_score[:,1] -= len(neg_excluded)

        # iterate over all prediction features of the dataset
        for feature in self.prediction_features:

            if sub_pos: # count unique values excluded

                pos_unique, pos_counts = np.unique(ex_pos_dataset[feature], return_counts=True)
                pos_occ = dict(zip(pos_unique, pos_counts))

                for value in pos_unique: # add the error that should not be substracted
                    atom_ind = self.atom_indexes[(feature, value)]
                    self.atom_score[atom_ind][0] += pos_occ[value]

            if sub_neg:

                neg_unique, neg_counts = np.unique(ex_neg_dataset[feature], return_counts=True)
                neg_occ = dict(zip(neg_unique, neg_counts))

                for value in neg_unique: # add the error that should not be substracted
                    atom_ind = self.atom_indexes[(feature, value)]
                    self.atom_score[atom_ind][1] += neg_occ[value]

        if sub_pos:
            del ex_pos_dataset

        if sub_neg:
            del ex_neg_dataset


        return

    #---------------------------------------------------------------------------
    def n_positives(self) -> int:
        return len(self._pos_samples)

    #---------------------------------------------------------------------------
    def n_negatives(self) -> int:
        return len(self._neg_samples)

    #---------------------------------------------------------------------------
    def n_atoms(self) -> int:
        return len(self.atoms)

    #---------------------------------------------------------------------------
    def get_atom(self, atom_index):
        return self.atoms[atom_index]

    #---------------------------------------------------------------------------
    def has_atom(self, atom):
        return atom in self.atoms

    #---------------------------------------------------------------------------
    def get_atom_index(self, atom):
        return self.atom_indexes[atom]

    #---------------------------------------------------------------------------
    def to_string(self):
        res = ''

        return res

    #---------------------------------------------------------------------------
    def compute_rule_error(self, rule, index) -> float:

        sample = self.dataset.loc[index]

        error = 0
        for atom_index in rule:
            atom = self.atoms[atom_index]
            if sample[atom[0]] != atom[1]:
                error += 1

        return error


    #---------------------------------------------------------------------------
    @staticmethod
    def create_instance(dataset, prediction_features, target_feature):

        # create an instance from a dataframe

        res = Instance() # create the new Instance object

        # associate a name and index to each atom
        res.atom_indexes = {} # dictionary of atom indexes
        res.atoms = [] # list of atoms: variable index and discrete value
        res.atom_score = [] # positive and negative score for each atom

        res._pos_samples = [] # positive samples of the instance
        res._neg_samples = [] # negative samples of the instance

        # initialization of atom attributes
        res.dataset = dataset
        res.prediction_features = prediction_features
        res.target_feature = target_feature

        res._init_instance() # initiallization of atom scores

        return res


    #---------------------------------------------------------------------------
    @staticmethod
    def create_sub_instance(main_instance, pos_excluded, neg_excluded = []):

        res = Instance()

        # only references are used here
        res.atoms = main_instance.atoms
        res.atom_indexes = main_instance.atom_indexes

        res.dataset = main_instance.dataset
        res.prediction_features = main_instance.prediction_features
        res.target_feature = main_instance.target_feature

        # copy are made for scores and sample classes
        res.atom_score = main_instance.atom_score.copy()

        res._pos_samples = [elt for elt in main_instance._pos_samples if not elt in pos_excluded]
        res._neg_samples = [elt for elt in main_instance._neg_samples if not elt in neg_excluded]

        res.n_values = main_instance.n_values

        res._init_sub_instance(pos_excluded, neg_excluded)

        return res

    #---------------------------------------------------------------------------
    @staticmethod
    def create_instance_explicit(dataset, pos_samples, neg_samples):

        # create the instance by explicitly stating positive and negative samples

        dataset['Class'] = False

        # initialize the class feature: positive and negative cells
        dataset.loc[pos_samples,'Class'] = True

        # creation of the associated classification instance
        prediction_features = dataset.columns.drop(['Class'])
        instance = Instance.create_instance(dataset, prediction_features, 'Class')

        return instance

    #---------------------------------------------------------------------------
    @staticmethod
    def create_cluster_instance(dataframe, clusters, label):

        # Join the dataframe with the cell labels
        dataframe = dataframe.join(clusters)

        dataframe['Class'] = (dataframe['Label'] == label)

        features = dataframe.columns
        prediction_features = features.drop(['Class', 'Label'])
        target_feature = 'Class'

        instance = Instance.create_instance(dataframe, prediction_features, target_feature)

        return instance

    #---------------------------------------------------------------------------
    @staticmethod
    def create_coexpression_instance(dataset, gene_label, value):

        dataset['Class'] = (dataset[gene_label] == value)

        features = dataset.columns
        prediction_features = features.drop(['Class', gene_label]) # the gene column is removed as well

        instance = Instance.create_instance(dataset, prediction_features, 'Class')

        return instance

    #---------------------------------------------------------------------------
    @staticmethod
    def create_regulation_instance(dataset, transitions, gene_label, value, ratio, pred_neq = 0):

        # pred_neq == 0: any predecessors
        # pred_neq == 1: only predecessors with a different value than the target value
        # pred_neq == 2: only predecessors with the same value

        # print('gene label: ', gene_label, ' ; value: ', value)
        # print('ratio: ', ratio)
        # print('pred neq: ', pred_neq)
        # print('n transitions: ', transitions.shape[0])

        dataset['Class'] = False

        positive_barcodes = []

        for barcode in dataset.index:

            # filter out the predecessors based on the target value
            if pred_neq == 0 or (pred_neq == 1 and dataset[gene_label][barcode] != value) or (pred_neq == 2 and dataset[gene_label][barcode] == value):

                # compute successors
                successors = (transitions.loc[transitions['T-1'] == barcode])['T']

                # compute the number of successors verifying the target value
                pos_suc = dataset.loc[successors,gene_label] == value

                if pos_suc.shape[0] > 0:

                    if ratio <= -0.5:
                        positive_barcodes.append(barcode)
                    else:
                        occ = pos_suc.value_counts()
                        suc_ratio = float(occ[True]) / float(pos_suc.shape[0])
                        if suc_ratio >= ratio:
                            positive_barcodes.append(barcode)


        # initialize the class feature: positive and negative cells
        dataset.loc[positive_barcodes,'Class'] = True

        # creation of the associated classification instance
        prediction_features = dataset.columns.drop(['Class'])
        instance = Instance.create_instance(dataset, prediction_features, 'Class')

        return instance

    #---------------------------------------------------------------------------
    @staticmethod
    def create_regulation_instance_delay(dataset, transitions, gene_label, value, ratio, delay, pred_neq = 0):

        print('test delay')

        # compute a graph on the transitions
        row_index = {}
        ind = 0
        for barcode in dataset.index.values:
            row_index[barcode] = ind
            ind += 1

        row_label = dataset.index.values.copy()

        # compute a graph of successors
        graph = [[] for _ in range(len(row_label))]
        for ind in range(len(row_label)):
            pred = row_label[ind]
            suc_labels = transitions.loc[transitions['T-1'] == pred]['T']
            for elt in suc_labels:
                graph[ind].append(row_index[elt])

        successors_delay = [[] for _ in range(len(row_label))] # list of successors at given distance

        # for each cell, compute successors at given distance (delay)
        for ind in range(len(row_label)):
            stack = [(ind,0)]
            while len(stack) > 0:
                top = stack.pop()
                if top[1] == delay:
                    successors_delay[ind].append(top[0])
                elif top[1] < delay:
                    for suc_ind in graph[top[0]]:
                        stack.append((suc_ind, top[1]+1))

        # compute the class vector based on the ratio
        dataset['Class'] = False
        for ind in range(len(row_label)):

            if pred_neq == 0 or (pred_neq == 1 and dataset[ind][gene_label] != value) or (pred_neq == 2 and dataset[ind][gene_label] == value):

                n_suc = len(successors_delay[ind])

                # compute the successors with the correct value
                suc_dataframe = dataset.iloc[successors_delay[ind],:]
                suc_dataframe = suc_dataframe[suc_dataframe[gene_label] == value]
                # print(suc_dataframe)

                if n_suc > 0:
                    if ratio <= -0.5 or suc_dataframe.shape[0]/n_suc >= ratio:
                        dataset.loc[row_label[ind],'Class'] = True

        # creation of the associated classification instance
        prediction_features = dataset.columns.drop(['Class'])
        instance = Instance.create_instance(dataset, prediction_features, 'Class')

        return instance


    #---------------------------------------------------------------------------
    @staticmethod
    def create_random_instance(dataset, proportion):

        # create a random instance from the dataset, having a given proportion of positive samples

        # don't forget the copy, otherwise, the cell order in the dataset is modified
        # as a result, the instance select the k first cells as positive -> well conditioned instance
        indexes = dataset.index.values.copy()
        # indexes = dataset.index.values

        np.random.shuffle(indexes)

        n_pos = int(proportion*float(len(indexes)))

        dataset['Class'] = False

        dataset.loc[indexes[:n_pos], 'Class'] = True

        prediction_features = dataset.columns.drop(['Class'])
        instance = Instance.create_instance(dataset, prediction_features, 'Class')

        return instance

File: c++/src/Instance.cpp
/*!
 * \file Instance.cpp
 * \author S. Buchet
 * \brief implementation of Instance class
 */


#include <fstream>

#include <sstream>

#include <algorithm>

#include "Instance.hpp"

#include "Combinatorics.hpp"

#include "Utils.hpp"

using namespace std;

/*----------------------------------------------------------------------------*/
Instance::Instance(const DataFrame<uint>& dataset, const vector<bool>& classVector):
_dataset(dataset)
{
  Combinatorics::generateRange(_dataset.nColumns(), _features);
  initialize(classVector);
}

/*----------------------------------------------------------------------------*/
Instance::Instance(const DataFrame<uint>& dataset, const vector<bool>& classVector, uint nval):
_dataset(dataset)
{
  Combinatorics::generateRange(_dataset.nColumns(), _features);
  initialize(classVector, nval);
}

/*----------------------------------------------------------------------------*/
Instance::Instance(const DataFrame<uint>& dataset, const std::vector<uint>& features, const std::vector<bool>& classVector, uint nval):
_dataset(dataset),
_features(features)
{
  initialize(classVector, nval);
}

/*----------------------------------------------------------------------------*/
Instance::Instance(const DataFrame<uint>& dataset, const std::vector<uint>& features, const std::vector<bool>& classVector):
_dataset(dataset),
_features(features)
{
  initialize(classVector);
}


/*----------------------------------------------------------------------------*/
Instance::Instance(const Instance& instance, const vector<uint>& excluded, uint nVal):
_dataset(instance._dataset)
{
  _features = instance._features;

  initializeSubInstance(instance, excluded, nVal);
}

/*----------------------------------------------------------------------------*/
void Instance::initialize(const vector<bool>& classVector) {

  for(uint index = 0; index < classVector.size(); index ++) {
    if(classVector[index]) {
      _positives.push_back(index);
    } else {
      _negatives.push_back(index);
    }
  }

  if(_positives.empty()) {
    cout << "WARNING: the instance should contain positive samples" << endl;
  }
  if(_negatives.empty()) {
    cout << "WARNING: the instance should contain negative samples" << endl;
  }

  /* number of variables */
  _nvar = _dataset.nColumns();

  /* init number of values for each variable */
  _nval.resize(_dataset.nColumns(), 0);

  /* indexing from the data */
  _posScore.resize(_dataset.nColumns());
  _negScore.resize(_dataset.nColumns());

  _atomIndexes.resize(_dataset.nColumns());


  /* lsit of counts for each value of each variable to compute the positive and negative scores */
  vector<vector<uint>> posCounts, negCounts;
  /* counting occurences in positive and negative samples */
  _dataset.uniqueCount(classVector, posCounts, negCounts);


  // for(uint indCol = 0; indCol < _dataset.nColumns(); indCol ++) {
  //
  //   /* initialization of the values and scores */
  //   _nval[indCol] = _dataset.nUnique(indCol);
  //
  //   _posScore[indCol].resize(_nval[indCol], 0);
  //   _negScore[indCol].resize(_nval[indCol], 0);
  //
  //   _atomIndexes[indCol].resize(_nval[indCol]);
  //
  //   for(uint indVal = 0; indVal < _nval[indCol]; indVal ++) {
  //
  //     /* create the list of atoms */
  //     _atoms.push_back(pair<uint,uint>(indCol, indVal));
  //     _atomIndexes[indCol][indVal] = static_cast<uint>(_atoms.size())-1;
  //
  //     /* initialization of the score for this atom */
  //     _posScore[indCol][indVal] = nPositives()-posCounts[indCol][indVal];
  //     _negScore[indCol][indVal] = nNegatives()-negCounts[indCol][indVal];
  //
  //   }
  //
  // }

  /* only use prediction features as atoms */
  for(auto& indCol : _features) {

    /* initialization of the values and scores */
    _nval[indCol] = _dataset.nUnique(indCol);

    _posScore[indCol].resize(_nval[indCol], 0);
    _negScore[indCol].resize(_nval[indCol], 0);

    _atomIndexes[indCol].resize(_nval[indCol]);

    for(uint indVal = 0; indVal < _nval[indCol]; indVal ++) {

      /* create the list of atoms */
      _atoms.push_back(pair<uint,uint>(indCol, indVal));
      _atomIndexes[indCol][indVal] = static_cast<uint>(_atoms.size())-1;

      /* initialization of the score for this atom */
      _posScore[indCol][indVal] = nPositives()-posCounts[indCol][indVal];
      _negScore[indCol][indVal] = nNegatives()-negCounts[indCol][indVal];

    }

  }

}

/*----------------------------------------------------------------------------*/
void Instance::initialize(const vector<bool>& classVector, uint nval) {

  /* _features.empty() means that all features are used */

  for(uint index = 0; index < classVector.size(); index ++) {
    if(classVector[index]) {
      _positives.push_back(index);
    } else {
      _negatives.push_back(index);
    }
  }

  if(_positives.empty()) {
    cout << "WARNING: the instance should contain positive samples" << endl;
  }
  if(_negatives.empty()) {
    cout << "WARNING: the instance should contain negative samples" << endl;
  }

  /* number of variables: if _features is empty, all variables are used */
  _nvar = _dataset.nColumns();
  _nval.resize(_dataset.nColumns(), nval);

  _atoms.resize(_nvar*nval);

  /* positive and negative scores for the atoms: size = dataset size but depending on the features everything is not used */
  _posScore.resize(_dataset.nColumns(), vector<uint>(nval, 0));
  _negScore.resize(_dataset.nColumns(), vector<uint>(nval, 0));

  /* same here: some atoms do not exist -> index=0 */
  _atomIndexes.resize(_dataset.nColumns(), vector<uint>(nval, 0));

  if(_features.size() == _nvar) {

    /* counting occurences in positive and negative samples */
    _dataset.uniqueCount(classVector, _posScore, _negScore);

    /* create all atoms of the instance */
    uint indAtom = 0;
    for(uint indCol = 0; indCol < _dataset.nColumns(); indCol ++) {
      for(uint indVal = 0; indVal < nval; indVal ++) {
        /* create the list of atoms */
        _atoms[indAtom] = pair<uint,uint>(indCol, indVal);
        _atomIndexes[indCol][indVal] = indAtom;
        indAtom += 1;

        /* initialization of the score for this atom */
        _posScore[indCol][indVal] = nPositives()-_posScore[indCol][indVal];
        _negScore[indCol][indVal] = nNegatives()-_negScore[indCol][indVal];
      }
    }

  } else { /* only a subset of the features are used */

    _dataset.uniqueCount(_features, classVector, _posScore, _negScore);

    /* creation of the atoms: only in the features list */
    uint indAtom = 0;
    for(uint& indCol: _features) {
      for(uint indVal = 0; indVal < nval; indVal ++) {
        /* create the list of atoms */
        _atoms[indAtom] = pair<uint,uint>(indCol, indVal);
        _atomIndexes[indCol][indVal] = indAtom;
        indAtom += 1;

        /* initialization of the score for this atom */
        _posScore[indCol][indVal] = nPositives()-_posScore[indCol][indVal];
        _negScore[indCol][indVal] = nNegatives()-_negScore[indCol][indVal];
      }
    }

  }

}

/*----------------------------------------------------------------------------*/
void Instance::initializeSubInstance(const Instance& instance, const vector<uint>& excluded, uint nValues) {

  // cout << "sub instance creation: " << endl;

  _nvar = instance._nvar;
  _nval = instance._nval;

  /* list of atoms are copied (same prediction features) */
  _atoms = instance._atoms;
  _atomIndexes = instance._atomIndexes;

  _features = instance._features;

  /* remove excluded positive samples */
  for(uint ind = 0; ind < instance._positives.size(); ind ++) {
    auto it = find(excluded.begin(), excluded.end(), ind);
    if(it == excluded.end()) {
      _positives.push_back(instance._positives[ind]);
    }
  }


  _negatives = instance._negatives;

  _posScore = instance._posScore;
  _negScore = instance._negScore;

  // vector<vector<uint>> counts(_nvar, vector<uint>(nValues,0));
  vector<vector<uint>> counts(_dataset.nColumns(), vector<uint>(nValues,0));

  /* list of sample indexes to exclude */
  vector<uint> excludedInd(excluded.size());
  for(uint ind = 0; ind < excluded.size(); ind ++) {
    excludedInd[ind] = instance._positives[excluded[ind]];
  }

  /* compute atom occurences in excluded positive samples */
  _dataset.uniqueCount(excludedInd, counts);

  if(_features.size() == _nvar) { /* all features of the dataset are used */

    /* compute the new scores considering the occurences to remove */
    for(uint indCol = 0; indCol < _nvar; indCol ++) {
      for(uint indVal = 0; indVal < nValues; indVal ++) {
        /* number of samples where value != indVal */
        uint nDiff = static_cast<uint>(excludedInd.size())-counts[indCol][indVal];
        _posScore[indCol][indVal] -= nDiff;
      }
    }

  } else { /* only a subset of the features are used in the dataset */

    /* compute the new scores considering the occurences to remove */
    for(uint& indCol: _features) {
      for(uint indVal = 0; indVal < nValues; indVal ++) {
        /* number of samples where value != indVal */
        uint nDiff = static_cast<uint>(excludedInd.size())-counts[indCol][indVal];
        _posScore[indCol][indVal] -= nDiff;

      }
    }

  }


  // cout << "n excluded: " << excluded.size() << endl;
  // cout << "original n positives: " << instance._positives.size() << endl;
  // cout << "n positives: " << _positives.size() << endl;


}

/*----------------------------------------------------------------------------*/
Instance::~Instance() {

}

/*----------------------------------------------------------------------------*/
uint Instance::getAtomIndex(Atom atom) const {
  return _atomIndexes[atom.first][atom.second];
}

/*----------------------------------------------------------------------------*/
Atom Instance::getAtom(uint index) const {
  return _atoms[index];
}

/*----------------------------------------------------------------------------*/
pair<std::string, uint> Instance::getAtomLabel(uint index) const {
  return pair<string, uint>(_dataset.getColLabel(_atoms[index].first), _atoms[index].second);
}


/*----------------------------------------------------------------------------*/
Score Instance::getAtomScore(uint index) const {
  return getAtomScore(_atoms[index]);
}

/*----------------------------------------------------------------------------*/
Score Instance::getAtomScore(Atom atom) const {
  return Score(_posScore[atom.first][atom.second], _negScore[atom.first][atom.second]);
}

/*----------------------------------------------------------------------------*/
uint Instance::getAtomPosScore(uint index) const {
  auto atom = _atoms[index];
  return getAtomPosScore(atom);
}

/*----------------------------------------------------------------------------*/
uint Instance::getAtomPosScore(Atom& atom) const {
  return _posScore[atom.first][atom.second];
}

/*----------------------------------------------------------------------------*/
uint Instance::getAtomNegScore(uint index) const {
  auto atom = _atoms[index];
  return getAtomNegScore(atom);
}

/*----------------------------------------------------------------------------*/
uint Instance::getAtomNegScore(Atom& atom) const {
  return _negScore[atom.first][atom.second];
}

/*----------------------------------------------------------------------------*/
uint Instance::matchingScore(const Body& body, uint indSample) const {
  uint score = 0;
  auto& row = _dataset.getRow(indSample);
  for(auto& ind : body) {
    if(row[_atoms[ind].first] != _atoms[ind].second) {
      score += 1;
    }
  }
  return score;
}

/*----------------------------------------------------------------------------*/
uint Instance::positiveMatchingScore(const Body& body, uint indPositive) const {
  return matchingScore(body, _positives[indPositive]);
}

/*----------------------------------------------------------------------------*/
uint Instance::negativeMatchingScore(const Body& body, uint indNegative) const {
  return matchingScore(body, _negatives[indNegative]);
}

/*----------------------------------------------------------------------------*/
const Sample& Instance::getSample(uint index) const {
  return _dataset.getRow(index);
}

/*----------------------------------------------------------------------------*/
const Sample& Instance::getPositive(uint ind) const {
  return _dataset.getRow(_positives[ind]);
}

/*----------------------------------------------------------------------------*/
const Sample& Instance::getNegative(uint ind) const {
  return _dataset.getRow(_negatives[ind]);
}

/*----------------------------------------------------------------------------*/
uint Instance::getPosClassIndex(uint indPos) const {
  return _positives[indPos];
}

/*----------------------------------------------------------------------------*/
uint Instance::getNegClassIndex(uint indNeg) const {
  return _negatives[indNeg];
}

/*----------------------------------------------------------------------------*/
uint Instance::nSamples() const {
  return _dataset.nRows();
}


/*----------------------------------------------------------------------------*/
uint Instance::nPositives() const {
  return static_cast<uint>(_positives.size());
}

/*----------------------------------------------------------------------------*/
uint Instance::nNegatives() const {
  return static_cast<uint>(_negatives.size());
}

/*----------------------------------------------------------------------------*/
uint Instance::nVariables() const {
  return _nvar;
}

/*----------------------------------------------------------------------------*/
uint Instance::nValues(uint varInd) const {
  return _nval.at(varInd);
}

/*----------------------------------------------------------------------------*/
uint Instance::nAtoms() const {
  return static_cast<uint>(_atoms.size());
}

/*----------------------------------------------------------------------------*/
uint Instance::nFeatures() const {
  return static_cast<uint>(_features.size());
}

/*----------------------------------------------------------------------------*/
const vector<uint>& Instance::getFeatures() const {
  return _features;
}

/*----------------------------------------------------------------------------*/
const DataFrame<uint>& Instance::getDataset() const {
  return _dataset;
}


/*----------------------------------------------------------------------------*/
string Instance::toString() const {

  string res = "";

  res += "n variables: " + to_string(_nvar) + "\n";
  res += "n atoms: " + to_string(_atoms.size()) + "\n";
  res += "n positive samples: " + to_string(_positives.size()) + "\n";
  res += "n negative samples: " + to_string(_negatives.size()) + "\n";

  return res;
}

/*----------------------------------------------------------------------------*/
uint Instance::initRandomClass(double proportion, vector<bool>& classVector) {

  uint nPos = static_cast<uint>(proportion*static_cast<double>(classVector.size()));

  /* make sure there are positive and negative samples */
  if(nPos == classVector.size()) {
    nPos --;
  } else if(nPos == 0) {
    nPos ++;
  }

  vector<uint> indexes;
  Combinatorics::generateRange(static_cast<uint>(classVector.size()), indexes);
  Utils::getInstance().shuffle(indexes);

  for(uint ind = 0; ind < nPos; ind ++) {
    classVector[indexes[ind]] = true;
  }

  return nPos;
}

/*----------------------------------------------------------------------------*/
uint Instance::initClusterClass(DataFrame<uint>& dataset, DataFrame<string>& labels, string cellLabel, vector<bool>& classVector) {

  uint nPos = 0;

  for(uint indRow = 0; indRow < labels.nRows(); indRow ++) {
    string label = labels.getRowLabel(indRow);

    if(labels.getData(indRow, 0) == cellLabel && dataset.containsRowIndex(label)) {
      classVector[dataset.getRowIndex(label)] = true;
      nPos ++;
    }

  }

  return nPos;
}

/*----------------------------------------------------------------------------*/
uint Instance::initCoexprClass(DataFrame<uint>& dataset, uint geneIndex, uint value, vector<uint>& features, vector<bool>& classVector) {

  uint nPos = 0;

  /* class vector: true iif gene value is the target value */
  for(uint indRow = 0; indRow < classVector.size(); indRow ++) {
    auto& sample = dataset.getRow(indRow);
    if(sample[geneIndex] == value) {
      classVector[indRow] = true;
      nPos ++;
    }
  }

  /* list of features: all features except targetFeature */
  features.resize(dataset.nColumns()-1);
  uint indFeature = 0;
  for(uint indCol = 0; indCol < dataset.nColumns(); indCol ++) {
    if(indCol != geneIndex) {
      features[indFeature] = indCol;
      indFeature ++;
    }
  }

  return nPos;
}

/*----------------------------------------------------------------------------*/
uint Instance::initRegulationClass(DataFrame<uint>& dataset, NGraph& graph, uint geneIndex, uint value, double proportion, uint predNeq, vector<bool>& classVector) {

  uint nPos = 0;

  for(uint indCell = 0; indCell < dataset.nRows(); indCell ++) {

    if(!graph[indCell].empty() && (predNeq == 0 || (predNeq == 1 && dataset.getRow(indCell)[geneIndex] != value) || (predNeq == 2 && dataset.getRow(indCell)[geneIndex] == value) ) ) { /* look at all successors */

      uint occurences = 0;
      for(auto& indSuc : graph[indCell]) {
        if(dataset.getRow(indSuc)[geneIndex] == value) {
          occurences ++;
        }
      }



      /* if the target atom is detected in enough successors, the sample is considered as positive */
      if( (proportion <= -0.5 and occurences > 0) || (static_cast<double>(occurences) / static_cast<double>(graph[indCell].size()) >= proportion) ) {
        classVector[indCell] = true;
        nPos ++;
      }

      // if(occurences > 0) {
      //   classVector[indCell] = true;
      //   nPos ++;
      // }

    }
  }

  return nPos;

}

/*----------------------------------------------------------------------------*/
ostream& operator<<(ostream& str, const Instance& instance) {
  str << instance.toString();
  return str;
}


/*----------------------------------------------------------------------------*/
ostream& operator<<(ostream& str, const pair<uint,uint>& tuple) {
  str << "(" << tuple.first << "," << tuple.second << ")";
  return str;
}

/*----------------------------------------------------------------------------*/
ostream& operator<<(ostream& str, const Body& body) {
  str << "[";
  for(uint ind = 0; ind < body.size(); ind ++) {
    str << body[ind];
    if(ind < body.size()-1) {
      str << ", ";
    }
  }
  str << "]";
  return str;
}

/*----------------------------------------------------------------------------*/
string bodyToString(const Body& body) {
  ostringstream str;
  str << body;
  return str.str();
}
